### PR TITLE
BaseSearchViewController

### PR DESCRIPTION
**Pull request by ArturAzarau**

### CODE REVIEW

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    private func viewModels(from items: ItemsList) -> [ItemViewModel] {
+        return items.map { self.viewModel(from: $0) }
+    }
+
+    var loadingResultObservable: Observable<ResultType> {
+        return loadingStateDriver
+            .asObservable()
+            .map { state -> ResultType? in
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewModel.swift]: map { $0.result }

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                } else {
+                    return .empty()
+                }
+            })
+    }
+
+    var loadingErrorObservable: Observable<Error> {
+        return loadingStateDriver
+            .asObservable()
+            .map { state -> Error? in
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewModel.swift]: map { $0.error }

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                    return .just(value)
+                } else {
+                    return .empty()
+                }
+            })
+            .share(replay: 1, scope: .forever)
+            .asDriver(onErrorDriveWith: .empty())
+    }
+
+    func viewModel(from item: Item) -> ItemViewModel {
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewModel.swift]: open

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+import TableKit
+import RxSwift
+
+open class BaseSearchViewController<Item,
+    ItemViewModel,
+    ViewModel,
+    ResultsVC: SearchResultsViewController,
+    CustomView>: UIViewController, ConfigurableController
+where ViewModel: BaseSearchViewModel<Item, ItemViewModel>, CustomView: UIView, CustomView: TableViewHolder {
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewController.swift]:
```swift
CustomView: UIView & TableViewHolder
```

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+    var searchBarPlaceholder: String {
+        return ""
+    }
+
+    var searchBarColor: UIColor {
+        return .gray
+    }
+
+    var statusBarView: UIView {
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+//  THE SOFTWARE.
+//
+
+import TableKit
+import RxSwift
+
+open class BaseSearchViewController<Item,
+    ItemViewModel,
+    ViewModel,
+    ResultsVC: SearchResultsViewController,
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+    // MARK: - Life Cycle
+
+    open override func loadView() {
+        view = customView
+    }
+
+    // MARK: - Configurable Controller
+
+    public func configureBarButtons() {
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+            .do(onNext: { [weak self] text in
+                self?.handle(searchText: text)
+            })
+            .map { $0 ?? "" }
+
+        viewModel.bind(searchText: searchText)
+            .disposed(by: disposeBag)
+    }
+
+    public func addViews() {
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        searchController.view.addSubview(statusBarView)
+    }
+
+    public func configureAppearance() {
+        definesPresentationContext = true
+        customView.tableView.tableHeaderView?.backgroundColor = searchBarColor
+    }
+
+    public func localize() {
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+    public func addViews() {
+        if #available(iOS 11.0, *) {
+            navigationItem.searchController = searchController
+        } else {
+            customView.tableView.tableHeaderView = searchController.searchBar
+        }
+        searchController.view.addSubview(statusBarView)
+    }
+
+    public func configureAppearance() {
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        customView.tableView.tableHeaderView?.backgroundColor = searchBarColor
+    }
+
+    public func localize() {
+        searchController.searchBar.placeholder = searchBarPlaceholder
+    }
+
+    // MARK: - Search Controller Functionality
+
+    func createRows(from itemsViewModels: [ItemViewModel]) -> [Row] {
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        fatalError("createRows(from:) has not been implemented")
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+    func handle(searchResultsState state: SearchResultsViewControllerState) {
+        searchResultsViewController.update(from: state)
+    }
+
+    func handle(searchText: String?) {
+        setTableViewInsets()
+    }
+
+    private func setTableViewInsets() {
+        if !didEnterText {
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Protocols/Controllers/SearchResultsViewController.swift [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+public protocol SearchResultsViewController where Self: UIViewController {
+
+    var resultsView: TableViewHolder? { get set }
+
+    func update(from state: SearchResultsViewControllerState)
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Protocols/Controllers/SearchResultsViewController.swift]: fixed

File: Sources/Protocols/Controllers/SearchResultsViewController.swift [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+public protocol SearchResultsViewController where Self: UIViewController {
+
+    var resultsView: TableViewHolder? { get set }
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Protocols/Controllers/SearchResultsViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+import RxCocoa
+
+open class BaseSearchViewModel<Item, ItemViewModel>: GeneralDataLoadingViewModel<[Item]> {
+
+    typealias ItemsList = [Item]
+
+    private let searchTextRelay = BehaviorRelay(value: "")
+
+    init(dataSource: Single<ItemsList>) {
+        super.init(dataSource: dataSource, emptyResultChecker: { _ in false })
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    open var itemsViewModelsDriver: Driver<[ItemViewModel]> {
+        return loadingResultObservable
+            .map { [weak self] items in
+                self?.viewModels(from: items)
+            }
+            .flatMap({ element -> Observable<[ItemViewModel]> in
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    open var searchResultsDriver: Driver<[ItemViewModel]> {
+        return searchTextRelay.throttle(1, scheduler: MainScheduler.instance)
+            .withLatestFrom(loadingResultObservable) { ($0, $1) }
+            .flatMapLatest { [weak self] searchText, items -> Observable<ItemsList> in
+                self?.search(by: searchText, from: items).asObservable() ?? .empty()
+            }
+            .map { [weak self] items in
+                self?.viewModels(from: items)
+            }
+            .flatMap({ element -> Observable<[ItemViewModel]> in
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+        return loadingStateDriver
+            .asObservable()
+            .map { state -> ResultType? in
+                guard case let .result(data, _) = state else {
+                    return nil
+                }
+
+                return data
+            }
+            .flatMap({ element -> Observable<ResultType> in
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+        return loadingStateDriver
+            .asObservable()
+            .map { state -> Error? in
+                guard case let .error(error) = state else {
+                    return nil
+                }
+
+                return error
+            }
+            .flatMap({ element -> Observable<Error> in
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewModel.swift (hunk 1 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewModel.swift (hunk 2 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewModel.swift (hunk 3 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift (hunk 4 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+//
+
+import TableKit
+import RxSwift
+
+open class BaseSearchViewController<Item,
+    ItemViewModel,
+    ViewModel,
+    ResultsVC: SearchResultsViewController,
+    CustomView>: UIViewController, ConfigurableController
```
[the remark this answers is not in the document]
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+open class BaseSearchViewController<Item,
+    ItemViewModel,
+    ViewModel,
+    CustomView: UIView & TableViewHolder>: BaseCustomViewController<ViewModel, CustomView>
+where ViewModel: BaseSearchViewModel<Item, ItemViewModel> {
+
+    // MARK: - Properties
+
+    private let disposeBag = DisposeBag()
+    private let searchResultsViewController: UIViewController & SearchResultsViewController
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewController.swift]: typealias it

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+    private lazy var searchController = UISearchController(searchResultsController: searchResultsViewController)
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewController.swift]: can be let

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+    // MARK: - Life Cycle
+
+    open override func loadView() {
+        view = customView
+    }
+
+    // MARK: - Configurable Controller
+
+    open override func configureBarButtons() {
+        // override in subclass
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewController.swift]: call super

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 25]
```diff
@@ -0,0 +1,190 @@
...
+        searchController.view.addSubview(statusBarView)
+    }
+
+    open override func configureAppearance() {
+        definesPresentationContext = true
+        customView.tableView.tableHeaderView?.backgroundColor = searchBarColor
+    }
+
+    open override func localize() {
+        searchController.searchBar.placeholder = searchBarPlaceholder
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewController.swift]: call super

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+        if #available(iOS 11.0, *) {
+            navigationItem.searchController = searchController
+        } else {
+            customView.tableView.tableHeaderView = searchController.searchBar
+        }
+        searchController.view.addSubview(statusBarView)
+    }
+
+    open override func configureAppearance() {
+        definesPresentationContext = true
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewController.swift]: call super

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 27]
```diff
@@ -0,0 +1,190 @@
...
+                self?.handle(searchText: text)
+            })
+            .map { $0 ?? "" }
+
+        viewModel.bind(searchText: searchText)
+            .disposed(by: disposeBag)
+    }
+
+    open override func addViews() {
+        if #available(iOS 11.0, *) {
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewController.swift]: super

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 28]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    // MARK: - Configurable Controller
+
+    open override func configureBarButtons() {
+        // override in subclass
+    }
+
+    open override func bindViews() {
+        viewModel.itemsViewModelsDriver
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewController.swift]: super

File: Sources/Classes/Search/BaseSearchViewController.swift [hunk 29]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    required public init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: - Life Cycle
+
+    open override func loadView() {
+        view = customView
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewController.swift]: BaseCustomViewController should do it for you

File: Sources/Classes/Search/BaseSearchViewController.swift (hunk 22 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift (hunk 23 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift (hunk 24 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift (hunk 29 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift (hunk 28 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift (hunk 27 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift (hunk 26 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewController.swift (hunk 25 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewController.swift]: fixed

File: Cartfile [hunk 30]
```diff
@@ -2,4 +2,5 @@ binary "https://raw.github.com/petropavel13/CarthageBinaries/master/SwiftDate/Sw
 github "ReactiveX/RxSwift"
 binary "https://raw.github.com/petropavel13/CarthageBinaries/master/RxAlamofire/RxAlamofire.json"
 github "maxsokolov/TableKit"
-github "pronebird/UIScrollView-InfiniteScroll"
\ No newline at end of file
+github "pronebird/UIScrollView-InfiniteScroll"
+GitHub "Alamofire/Alamofire" "4.8.1"
```
Comment (petropavel13) [inline Cartfile]: uncapitalize

File: Cartfile.resolved [hunk 31]
```diff
@@ -1,6 +1,6 @@
+binary "https://raw.github.com/petropavel13/CarthageBinaries/master/RxAlamofire/RxAlamofire.json" "4.3.0"
+binary "https://raw.github.com/petropavel13/CarthageBinaries/master/SwiftDate/SwiftDate.json" "5.1.0"
```
Comment (petropavel13) [inline Cartfile.resolved]: use TouchInstinct repo

File: LeadKit.xcodeproj/project.pbxproj [hunk 32]
```diff
@@ -2972,7 +3008,7 @@
 		67186B3A1EB248F100CFAFFB /* Debug */ = {
 			isa = XCBuildConfiguration;
 			buildSettings = {
-				APPLICATION_EXTENSION_API_ONLY = YES;
+				APPLICATION_EXTENSION_API_ONLY = NO;
```
Comment (petropavel13) [inline LeadKit.xcodeproj/project.pbxproj]: revert

Comment (petropavel13) [reply LeadKit.xcodeproj/project.pbxproj]: should be able to compile for extension as well

File: Sources/Classes/Search/BaseSearchViewController.swift (hunk 24 above)
Comment (petropavel13) [reply Sources/Classes/Search/BaseSearchViewController.swift]: method looks useless now

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 33]
```diff
@@ -0,0 +1,117 @@
...
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import RxSwift
+import RxCocoa
+
+private protocol OptionalType {
```
Comment (petropavel13) [inline Sources/Classes/Search/BaseSearchViewModel.swift]: move to separate file. file should start with BaseSearchViewModel definition

File: Cartfile (hunk 30 above)
Comment (ArturAzarau) [reply Cartfile]: fixed

File: Cartfile.resolved (hunk 31 above)
Comment (ArturAzarau) [reply Cartfile.resolved]: fixed

File: LeadKit.xcodeproj/project.pbxproj (hunk 32 above)
Comment (ArturAzarau) [reply LeadKit.xcodeproj/project.pbxproj]: But it shouldn't
image

File: Sources/Classes/Search/BaseSearchViewModel.swift (hunk 33 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: fixed

File: Sources/Protocols/Controllers/SearchResultsViewController.swift [hunk 34]
```diff
@@ -0,0 +1,30 @@
...
+//  THE SOFTWARE.
+//
+
+/// Protocol that represents a ViewController able to present search results
+public protocol SearchResultsViewController {
+
+    var searchResultsView: TableViewHolder { get set }
+
+    func update(for state: SearchResultsViewControllerState)
+
```
Comment (ZharaOo) [inline Sources/Protocols/Controllers/SearchResultsViewController.swift]: Remove line

Comment (ArturAzarau) [reply Sources/Protocols/Controllers/SearchResultsViewController.swift]: fixed

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 35]
```diff
@@ -0,0 +1,100 @@
...
+            .share(replay: 1, scope: .forever)
+            .asDriver(onErrorDriveWith: .empty())
+    }
+
+    open func viewModel(from item: Item) -> ItemViewModel {
+        fatalError("viewModel(from:) has not been implemented")
+    }
+
+    open func search(by searchString: String, from items: ItemsList) -> Single<ItemsList> {
+        fatalError("searchEngine(for:) has not been implemented")
```
Comment (iznv) [inline Sources/Classes/Search/BaseSearchViewModel.swift]: search(by:)?

File: Sources/Classes/Search/BaseSearchViewModel.swift [hunk 36]
```diff
@@ -0,0 +1,100 @@
...
+            .map { [weak self] items in
+                self?.viewModels(from: items)
+            }
+            .filterNil()
+            .share(replay: 1, scope: .forever)
+            .asDriver(onErrorDriveWith: .empty())
+    }
+
+    open var searchResultsDriver: Driver<[ItemViewModel]> {
+        return searchTextRelay.throttle(1, scheduler: MainScheduler.instance)
```
Comment (iznv) [inline Sources/Classes/Search/BaseSearchViewModel.swift]: why throttle, not debounce?

Comment (iznv) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: interval to variable (to be customized), 1 sec is too much, btw

File: Sources/Classes/Search/BaseSearchViewModel.swift (hunk 35 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: yes

File: Sources/Classes/Search/BaseSearchViewModel.swift (hunk 36 above)
Comment (ArturAzarau) [reply Sources/Classes/Search/BaseSearchViewModel.swift]: fixed

File: LeadKit.xcodeproj/project.pbxproj [hunk 37]
```diff
@@ -834,7 +842,9 @@
 			isa = PBXGroup;
 			children = (
 				6741CEC520E2434900FEC4D9 /* Controllers */,
+				72AECC68224A979D00D12E7C /* Search */,
 				6774527E2062566D0024EEEF /* DataLoading */,
+				72D213A222048162003B4292 /* Search */,
```
Comment (petropavel13) [inline LeadKit.xcodeproj/project.pbxproj]: twice?

File: LeadKit.xcodeproj/project.pbxproj [hunk 38]
```diff
@@ -892,7 +902,9 @@
 		671461D61EB3396E00EAB194 /* Enums */ = {
 			isa = PBXGroup;
 			children = (
+				72039CDE220899D000875DD4 /* Search */,
 				67274767206CCB6F00725163 /* Views */,
+				72AECC6D224A97B100D12E7C /* Search */,
```
Comment (petropavel13) [inline LeadKit.xcodeproj/project.pbxproj]: twice?

File: CHANGELOG.md [hunk 39]
```diff
@@ -1,4 +1,9 @@
 # Changelog
+### 0.9.15
+- **Add**: `BaseSearchViewController` class that that allows to enter text for search and then displays search results in table view.
```
Comment (petropavel13) [inline CHANGELOG.md]: that that

File: Sources/Protocols/OptionalType.swift [hunk 40]
```diff
@@ -0,0 +1,42 @@
...
+
+    var optional: Wrapped? { get }
+}
+
+extension Optional: OptionalType {
+    public var optional: Wrapped? { return self }
+}
+
+extension Observable where Element: OptionalType {
+    func filterNil() -> Observable<Element.Wrapped> {
```
Comment (petropavel13) [inline Sources/Protocols/OptionalType.swift]: unused in code, and internal for other users

File: LeadKit.xcodeproj/project.pbxproj (hunk 37 above)
Comment (ArturAzarau) [reply LeadKit.xcodeproj/project.pbxproj]: yes, twice

File: LeadKit.xcodeproj/project.pbxproj (hunk 38 above)
Comment (ArturAzarau) [reply LeadKit.xcodeproj/project.pbxproj]: above

File: CHANGELOG.md (hunk 39 above)
Comment (ArturAzarau) [reply CHANGELOG.md]: fixed

File: Sources/Protocols/OptionalType.swift (hunk 40 above)
Comment (ArturAzarau) [reply Sources/Protocols/OptionalType.swift]: deleted

**Pull request merged**